File: Java/Stack.java
public class Stack {
    private int capacity; // 스택의 최대 크기
    private int[] array; // 스택 데이터를 저장할 배열
    private int top; // 스택의 최상단을 나타내는 인덱스

    public Stack(int capacity){
        this.capacity = capacity;
        this.array = new int[capacity];
        this.top = -1; // 초기 상태: 스택이 비어있음
    }
    
    public boolean isEmpty(){
        return top == -1;
    }

    public boolean isFull(){
        return capacity -1 == top;
    }

    public void push(int value){
        if(!isFull()){
            top++;
            array[top] = value;
        }else{
            System.out.println("Stack Overflow");
        }
    }

    public int pop(){
        if(!isEmpty()){
            int value = array[top];
            top--;
            return value;
        }else{
            System.out.println("Stack Underflow");
            return -1;
        }
    }

    public int size(){
        return top + 1;
    }

    public void display(){
        for(int i = 0; i <= top; i++){
            System.out.println(array[i] + " ");
        }
        System.out.println();
    }


    public static void main(String[] args){
        Stack stack = new Stack(5);

        stack.push(10);
        stack.push(20);
        stack.push(30);

        System.out.println(" 스택 내용 : ");
        stack.display();

        System.err.println(stack.pop());
        System.out.println();
        stack.display();
        
        System.out.println( stack.size());
       
    }
}

File: Python/Syntax/Stack.py
# 클래스 선언
class Stack:
    def __init__(self, capacity): # 생성자 메서드
        self.capacity = capacity # 값 초기화
        self.array = [None] * self.capacity # 배열
        self.top = -1
        
    def isEmpty(self):
       return self.top == -1
   
    def isFull(self):
        return self.capacity - 1 == self.top
    
    def push(self, value):
        if not self.isFull():
            self.top += 1
            self.array[self.top] = value
        else:
            print("Stack Overflow")
    
    def pop(self):
        if not self.isEmpty():
            self.top -= 1
            return self.array[self.top + 1]
        else:
            print("Stack Underflow")
            
            
    def size(self):
        return self.top + 1
    
    def display(self):
        for i in range(self.top + 1):
            print(self.array[i], " ", end="") 
            


# 객체 생성
s1 = Stack(10)
s2 = Stack(100)

s1.push(10)
s1.push(20)
s1.push(30)


print(s1.pop())
print(s1.size())
s1.display()




# 괄호 짝맞추기 체크 함수
def check_brackets(str, length):
    s1 = Stack(length)
    
    for ch in str:
        if ch == "(":
            s1.push(ch)
        elif ch == ")":
            if s1.isEmpty():
                return False
            else:
                s1.pop()
        else:
            pass
    return s1.isEmpty()
            
        

str = input("수식 입력 : ")
print("괄호 짝 맞추기 결과 :" , check_brackets(str, len(str)))
      
